src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala

```diff
@@ -285,6 +292,8 @@ case class DeltaSource(
     val changes = if (start.isEmpty) {
       if (endOffset.isStartingVersion) {
         getChanges(endOffset.reservoirVersion, -1L, isStartingVersion = true)
+      } else if (options.startingVersion.isDefined || options.startingTimestamp.isDefined) {
+        getChanges(endOffset.reservoirVersion, -1L, isStartingVersion = false)
       } else {
         assert(endOffset.reservoirVersion > 0, s"invalid reservoirVersion in endOffset: $endOffset")
```

Comment:
IIUC, we are introducing a new call to getBatch like this (start is `None` and `isStartingVersion` in end is `false`):

```
getBatch(
  start = None,
  end = `{"sourceVersion":1,
          "reservoirId":"testId",
          "reservoirVersion":1,
          "index":-1,
          "isStartingVersion":false}`
  )
```

- If `reservoirVersion` is 0 (using `startingVersion = 0`), then the query running without this patch will fail.
- If `reservoirVersion` is greater than 0, then the query running without this patch will return a different answer: it returns all files in `getSnapshot(reservoirVersion)`.

Of cause, this is an issue only in the first batch and only when
- The offset log of the first batch is generated in a new Delta Lake version.
- The first batch fails.
- Rollback to the old version and rerun it. In this case, the  re-generated batch will be different than the previous run.

I think this should be okay. It's really hard to hit this issue. @marmbrus what do you think?

Reply:
I also do think it is really hard. I think we should document it, but I don't think it warrants a protocol change